drivers/gpio/gpio_bflb.c

```diff
@@ -0,0 +1,319 @@
+
+struct gpio_bflb_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	uint32_t base_reg;
+	void (*irq_config_func)(const struct device *dev);
+	void (*irq_enable_func)(const struct device *dev);
+};
+
+struct gpio_bflb_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	sys_slist_t callbacks;
+};
+
+static int gpio_bflb_port_get_raw(const struct device *dev, uint32_t *value)
+{
+	const struct gpio_bflb_config *const cfg = dev->config;
+
+	*value = sys_read32(cfg->base_reg + GLB_GPIO_CFGCTL30_OFFSET);
+
+	return 0;
+}
+
+static int gpio_bflb_port_set_masked_raw(const struct device *dev,
+					uint32_t mask,
+					uint32_t value)
+{
+	const struct gpio_bflb_config *const cfg = dev->config;
+	uint32_t tmp = 0;
+
+	tmp = sys_read32(cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+	tmp = (tmp & ~mask) | (mask & value);
+	sys_write32(tmp, cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+
+	return 0;
+}
+
+static int gpio_bflb_port_set_bits_raw(const struct device *dev, uint32_t mask)
+{
+	const struct gpio_bflb_config *const cfg = dev->config;
+	uint32_t tmp = 0;
+
+	tmp = sys_read32(cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+	tmp |= mask;
+	sys_write32(tmp, cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+
+	return 0;
+}
+
+static int gpio_bflb_port_clear_bits_raw(const struct device *dev, uint32_t mask)
+{
+	const struct gpio_bflb_config *const cfg = dev->config;
+	uint32_t tmp = 0;
+
+	tmp = sys_read32(cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+	tmp &= ~mask;
+	sys_write32(tmp, cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+
+	return 0;
+}
+
+static int gpio_bflb_port_toggle_bits(const struct device *dev, uint32_t mask)
+{
+	const struct gpio_bflb_config *const cfg = dev->config;
+	uint32_t tmp = 0;
+
+	tmp = sys_read32(cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+	tmp ^= mask;
+	sys_write32(tmp, cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+
+	return 0;
+}
+
+static void gpio_bflb_port_interrupt_configure_mode(const struct device *dev,
+						    uint32_t pin,
+						    enum gpio_int_mode mode,
+						    enum gpio_int_trig trig)
+{
+	const struct gpio_bflb_config *const cfg = dev->config;
+	uint32_t tmp = 0;
+	uint8_t trig_mode = 0;				/* default to 'sync' mode */
+
+	tmp = sys_read32(cfg->base_reg + GLB_GPIO_INT_MODE_SET1_OFFSET + ((pin / 10) << 2));
+	tmp &= ~(0x07 << ((pin % 10) * 3));		/* clear modes */
+
+	if ((trig & GPIO_INT_HIGH_1) != 0) {
+		trig_mode |= 1;
+	}
+
+	if (!(mode & GPIO_INT_EDGE)) {
+		trig_mode |= 2;
+	}
+
+	tmp |= (trig_mode << ((pin % 10) * 3));
+	sys_write32(tmp, cfg->base_reg + GLB_GPIO_INT_MODE_SET1_OFFSET + ((pin / 10) << 2));
+}
+
+static void gpio_bflb_pin_interrupt_clear(const struct device *dev, uint32_t mask)
+{
+	const struct gpio_bflb_config *const cfg = dev->config;
+
+	sys_write32(mask, cfg->base_reg + GLB_GPIO_INT_CLR1_OFFSET);
+
+	while ((sys_read32(cfg->base_reg + GLB_GPIO_INT_STAT1_OFFSET) & mask) != 0) {
+		__asm__ volatile ("nop");
+	}
```

Comment:
CC: @VynDragon ?

Reply:
This is supposed to have a timeout (about 32 loops of counting in SDK) yea, not sure why it doesnt.

Reply:
@VynDragon , could share the snip so I can replace to the correct usage?

Reply:
This is the version from the SDK, GLB_GPIO_INT0_CLEAR_TIMEOUT is 32
`/* timeout check */
            timeOut = GLB_GPIO_INT0_CLEAR_TIMEOUT;

            do {
                timeOut--;
            } while ((SET == GLB_Get_GPIO_IntStatus(gpioPin)) && timeOut);`

i suppose this would do
```c
static void gpio_bflb_pin_interrupt_clear(const struct device *dev, uint32_t mask)
{
	const struct gpio_bflb_config *const cfg = dev->config;
        uint8_t timeout = <somedefinenameyoulikewithvalue32>

	sys_write32(mask, cfg->base_reg + GLB_GPIO_INT_CLR1_OFFSET);

	while ((sys_read32(cfg->base_reg + GLB_GPIO_INT_STAT1_OFFSET) & mask) != 0
                 && timeout > 0) {
		timeout--;
	}

	sys_write32(0x0, cfg->base_reg + GLB_GPIO_INT_CLR1_OFFSET);
}```

Reply:
Fixed